### DISCUSSION TITLE
Question about how string.Format works

### DISCUSSION THREAD

**Discussion by zxcvqwerasdf** [dotnet/runtime#87512] [Q&A] [answered]
Environment: NET 6, Windows 10, No code optimization
Code:
```
using System.Diagnostics;
internal class Program
{
    public static void TestMethod()
    {
        for (int i = 0; i < 100000000; i++)
        {
            string tempTest = $"It's test: {i}";
        }
    }
    private static void Main(string[] args)
    {
        Stopwatch st = new Stopwatch();
        Thread.Sleep(1000);
        st.Restart();
        TestMethod();
        st.Stop();
        Console.WriteLine(st.Elapsed.ToString());
        Console.Read();
    }
}
```
Now replace {i} to {i.ToString()}, recompile and run.
For {i} the execution time is about 4.11 sec
For {i.ToString()} the execution time is about 2.6 sec

Why? I checked MSIL code and got this

{i} was changed to this
![image](https://github.com/dotnet/runtime/assets/67195196/1c5875f0-62b9-47ec-9e44-c64462a26a87)

and {i.ToString()} is just .ToString call and string.Format(string, string) proceed
![image](https://github.com/dotnet/runtime/assets/67195196/8ea9822b-f782-4974-840d-3035f8e52888)

i checked what happening in NET 5 and got this:
![image](https://github.com/dotnet/runtime/assets/67195196/6408d7f3-08c3-4ed1-ad84-595d393a2132)
It's boxing now, I checked the performance and and it was ~6 sec.

Why it's happening? Why {i.ToString()} is way better in performance vs {i}? 
Thank you.

**Comment by Clockwork-Muse** [comment]
> Why {i.ToString()} is way better in performance vs {i}?

The way this code is written, there could be a number of external factors influencing any alleged performance numbers.  We use [BenchmarkDotNet](https://benchmarkdotnet.org/) to mitigate these sorts of factors.
Please re-write your code using the library and update this issue as appropriate.

**Reply by zxcvqwerasdf** [reply]
![image](https://github.com/dotnet/runtime/assets/67195196/2a0abcae-2edd-419d-9f1a-e6c0fee92009)
```
namespace Benchmark
{
    public class StringInterpolationBench
    {
        [Benchmark]
        public void StringInterpolationWithoutToString()
        {
            for (int i = 0; i < 100000000; i++)
            {
                string tempTest = $"It's test: {i}";
            }
        }

        [Benchmark]
        public void StringInterpolationWithToString()
        {
            for (int i = 0; i < 100000000; i++)
            {
                string tempTest = $"It's test: {i.ToString()}";
            }
        }

    }
}
```

**Comment by gfoidl** [comment, marked as answer]
See [String Interpolation in C# 10 and .NET 6](https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/) for an in-depth write-out about that topic.

Also note that `i.ToString` creates a string (so an allocation), that is avoided with the plain string interpolation.
